### DISCUSSION TITLE
initQpStatefulProperties() was not overridden in the derived class

### DISCUSSION THREAD

**Discussion by ritamcj** [idaholab/moose#19450] [Q&A General] [answered]
Hi everyone,
I am simulating nucleation and growth for a polycrystal and trying to assign crystallographic orientations to new grains. I am attempting to read the updated orientation tensor from phase field into FiniteStrainUObasedCP in tensor mechanics. When I try input the "old" state property, I see the error as attached.

![Screenshot from 2021-11-19 19-13-40](https://user-images.githubusercontent.com/81842276/142877852-f3a4053c-9f6c-4ddc-8517-8456ca584b78.png)

 I came across this discussion https://groups.google.com/g/moose-users/c/96d0ZUsyJQ0 and understood the cause of this error. I then tried to initialize the tensor in initQpStatefulProperties() as `_crysrot_nuclei[_qp].setToIdentity();` but this didn't work as attached. I tried numerous other approaches but none worked.

![Screenshot from 2021-11-22 18-54-25](https://user-images.githubusercontent.com/81842276/142877201-f8964ca6-587c-4991-9594-f49b1c184531.png)

When I tried to read the current state property, I ran into "Cyclic dependency detected in object ordering:", possibly due to reading a variable calculated at "timestep_end" inside a function that is called at "timestep_begin".

![Screenshot from 2021-11-22 19-02-02](https://user-images.githubusercontent.com/81842276/142878534-0d6761de-0057-40ca-9d58-929f61816123.png)

Kindly let me know possible approaches to solve this problem. Any help is greatly appreciated.

Warm regards,
Ritam Chatterjee

**Comment by GiudGiud** [comment]
Hi Rittam

Have you looked through other examples in the code for doing this?
There are plenty in tensor_mechanics/src/materials.

Guillaume

**Reply by ritamcj** [reply]
Hi Guillaume,
Thank you for your response. I will search in the mentioned folder.
Ritam

**Reply by ritamcj** [reply]
Hi Guillaume,
I looked into various files such as Eshelbytensor.C, RankTwoInvariant.C etc., and tried various approaches but couldn't solve this error. Basically, if I declare a property in a file in phase field at the end of a time step, how do I import its value to a file in tensor mechanics at the beginning of the next time step? Hence, an old stateful property is to be imported.
Ritam

**Reply by GiudGiud** [reply]
To import a previous time step material property, you can use getMaterialPropertyOld<return type (Real, ADReal..., same as the one it was declared with)>("name_in_input_file OR name_of_the_property")

**Comment by hugary1995** [comment, marked as answer]
@ritamcj, when you do `_crysrot_nuclei[_qp].setToIdentity();` and you see error messages containing `'this' argument discards qualifiers`, you probably mistakenly declared the material property `_crysrot_nuclei` as `const` in the header file.

**Reply by ritamcj** [reply]
Hi Gary,
Thank you for your reply. You are right. However, when I remove `const` from the header file, I get the error as in the screenshot. I think `const` is required when importing an old state property. 

I basically want to import an old state property from the phase field module to ComputeElasticitytensorCP, which runs at the beginning of a time step. But, this calls `initQpStatefulProperties()` and tries to override it. I don't know how to solve this. I declared the property in phase field as `_crysrot_nuclei(declareProperty<RankTwoTensor>("crysrot_nuclei"))` and then read the old state property as `_crysrot_nuclei_old(getMaterialPropertyOld<RankTwoTensor>("crysrot_nuclei"))`. Do let me know any mistake in my approach.

Warm regards,
Ritam

![Screenshot from 2021-12-01 23-45-04](https://user-images.githubusercontent.com/81842276/144291113-e60edff9-930a-4c92-9f99-8456bc8903f5.png)

**Reply by loganharbour** [reply]
The type should be a const reference in this case.

**Reply by hugary1995** [reply]
The old property _crysrot_nuclei_old should be a const reference. The property itself _crysrot_nuclei that you declared should be a reference. If it still doesn't work, could you paste the relevant pieces of your constructor, initQpStatefulProperties call, and header file here?

There is very likely a minor mistake in your header that is easy to identify.

**Reply by ritamcj** [reply]
Hi Gary, 
I will take care to follow the guidelines next time. I have attached some pseudo code that will help understand the issue and my approach. I am trying to declare a property called `Euler_angles_nuclei` in Euler_nuclei.C in phase field and am trying to read it's old state value in elasticity_tensor.C in tensor mechanics. I have not explicitly called `initQpStatefulProperties` in elasticity_tensor.C. It is being called inside FiniteStrainUObasedCP in tensor mechanics. I hope the example files are easily understandable.
Warm regards,
Ritam

[Euler_nuclei_Cfile.txt](https://github.com/idaholab/moose/files/7644022/Euler_nuclei_Cfile.txt)
[Euler_nuclei_hfile.txt](https://github.com/idaholab/moose/files/7644023/Euler_nuclei_hfile.txt)
[elasticity_tensor_Cfile.txt](https://github.com/idaholab/moose/files/7644024/elasticity_tensor_Cfile.txt)
[elasticity_tensor_hfile.txt](https://github.com/idaholab/moose/files/7644025/elasticity_tensor_hfile.txt)

**Reply by hugary1995** [reply]
The framework is designed in the following way:
1. Only the object that declares the material property can assign values to it.
2. Objects can request material properties, gaining read-only access to their values.
3. When any object (including the object that declares it) requests the old value of a material property, that property becomes "stateful".
4. All stateful material properties must be initialized within the `initQpStatefulProperties` call.  Due to rules 1 & 2, you can only do this in the object that declares the material property.

**Reply by hugary1995** [reply]
By read-only, I mean `const MaterialProperty<> &` -- you can't write to a const reference.

Another issue I noticed is that you are mixing material property types. Your `Euler_angles_nuclei` was declared as `RealVectorValue`s in `Euler_nuclei.h` but is retrieved as `RankTwoTensor`s in `elasticity_tensor.C`. If the types don't match, the framework cannot locate the material property you request.

**Reply by ritamcj** [reply]
Hi Gary,
Thank you for the detailed explanation, especially point 4. I now understand that I'll have to try a different approach. I mistakenly retrieved the property as a rank two tensor while setting up the pseudo code. Many thanks for your suggestions.
Warm regards,
Ritam

**Comment by hugary1995** [comment]
Please also follow the posting guidelines #18270 -- don't upload screenshotd, instead use the native markdown code formatting.
